### ISSUE TITLE
cli: Add --scaffold flag to `anchor init` command

### ISSUE THREAD

**Issue by armaniferrante** [project-serum/anchor#480]
In addition to initializing the workspace, the `--scaffold` flag should add a batteries included react app to the workspace with a wallet preconfigured.

Bonus: hot reloading so that whenever the smart contract changes, it can be upgraded and the UI can interact with the upgraded program automatically.

For inspiration, see the excellent https://github.com/austintgriffith/scaffold-eth#-scaffold-eth.

**Comment by colinlgray** [comment]
I looked into this a bit and I think that adding a `--scaffold` flag might not be what you want. The way that [scaffold-eth](https://github.com/austintgriffith/scaffold-eth) is structured is that there is the [full react app](https://github.com/austintgriffith/scaffold-eth/tree/master/packages/react-app) inside one of the folders of repo. If you wanted to create a react app that was specific to anchor, you would probably put the source for that app in this repo, but then there would be no reason to do anything on init because the files that it would create already exist inside the repo. The user would just move to the folder, `yarn install`, and `yarn start`

  If you wanted to use a `--scaffold` flag and not put a react app into this repo, you could do something like have the init command create a directory and clone [dapp-scaffold](https://github.com/solana-labs/dapp-scaffold) into it. The problem with that is it would be lacking any knowledge of anchor and you couldn't add features easily.

If you're looking to do something like eth scaffold, and create a new react app that is set up for anchor I might be able to help out.

**Comment by colinlgray** [comment]
I was thinking about this more and I got an idea about how this could work. What I would recommend doing is forking [Create React App](https://github.com/facebook/create-react-app) and adding in the [hooks](https://github.com/solana-labs/dapp-scaffold/tree/main/src/hooks) from dapp-scaffold, some basic components, and any other changes specific to anchor. Then the `anchor init --scaffold` would create a directory and clone that database.

What's nice about this is that whenever CRA is updated by the React team to add a feature, that can be rebased into the forked repo. That way it will always have the most up to date React development environment.

You will probably want that repo to be owned by project serum but I can take a first shot at what that would look like

**Comment by colinlgray** [comment]
Another update on this. The right thing to do here is create a [custom Create React App template](https://create-react-app.dev/docs/custom-templates/). Then on anchor init it will run 

```
npx create-react-app my-app --template cra-template-solana
```


I've started on a template for it [here](https://github.com/colinlgray/cra-template-solana). Right now it's a copy/paste of [dapp-scaffold](https://github.com/solana-labs/dapp-scaffold). It has a couple issues but it basically works. I'm gonna keep messing with it

**Comment by armaniferrante** [comment]
Awesome. Will this auto update when create react app updates?

Can we remove antd and all the .less files? Id prefer to be agnostic to styling frameworks.

**Comment by colinlgray** [comment]
Yeah so the way this works is that create-react-app creates like a "wrapped" react app that has scripts, build tools, etc and until you run `yarn eject` on the app, it stays in the system and gets updated with the latest and greatest. Using the template specifies what source files start out in that app. When I was doing this I realized that dapp scaffold is a CRA app as well. It uses [craco](https://github.com/gsoft-inc/craco) to manage that. So this is basically the same as cloning dapp scaffold but using the CRA cli.

Right now this thing is a copy/paste of dapp scaffold so that's why it uses ant and less. No styling frameworks would mean new css right? What should this look like?

**Comment by colinlgray** [comment]
I've been thinking about this some more and I want to give an argument for leaving antd in.

I think that when the user uses this template and creates an app, what they want is to focus on their product and build. They want to put some buttons on a screen, connect it to their contract, and iterate. The design library does provide styling, but it also contains components needed to build an app like [Button](https://ant.design/components/button/) and [DatePicker](https://ant.design/components/date-picker/). By removing a design library we would be doing a disservice to the users because then they would be forced to either build their own base components, or choose a framework themselves. It is completely possible for me to build an app and not use a design framework, however that is effectively me writing my own framework and honestly any that I build would not be as robust as antd. Additionally this is all configurable to the user. If they want to remove antd and replace it with something else that is not difficult.

**Comment by armaniferrante** [comment]
I imagine there would be different templates for different preferences. The base template would have no framework and then others would build on top of that.

**Comment by colinlgray** [comment]
What would that look like with no framework? Like default html?

**Comment by colinlgray** [comment]
I think it's reasonable to have a lot of templates, each with a different design library or tooling, however I don't think that should be the default template, or the first template built. My worry is that someone would use that template without a design library, not really understand web development, and continue not using one. In my opinion the base experience for scaffolding a web app should be the one where the user doesn't have to make any decisions about any build processes or dependencies, and can start immediately building their product. Including a design framework is the best way to do that

**Comment by WrRaThY** [comment]
hey. maybe not the perfect place for that question, but...
is there any example of how rust tests (not JS) would work with anchor?

I'm asking, because in the dapp-scaffold there is an integration test (or a better one can be found here: https://github.com/solana-labs/example-helloworld) 
```
#[tokio::test]
async fn test_helloworld() {
    let program_id = Pubkey::new_unique();
    let greeted_pubkey = Pubkey::new_unique();

    let mut program_test = ProgramTest::new(
        "helloworld", // Run the BPF version with `cargo test-bpf`
        program_id,
        processor!(process_instruction), // Run the native version with `cargo test`
    );
```

but... this is nowhere to be found in case of anchor
```
        processor!(process_instruction), // Run the native version with `cargo test`
```

any ideas how to work with that?

**Comment by armaniferrante** [comment]
> hey. maybe not the perfect place for that question, but...
> is there any example of how rust tests (not JS) would work with anchor?
> 
> I'm asking, because in the dapp-scaffold there is an integration test (or a better one can be found here: https://github.com/solana-labs/example-helloworld)
> 
> ```
> #[tokio::test]
> async fn test_helloworld() {
>     let program_id = Pubkey::new_unique();
>     let greeted_pubkey = Pubkey::new_unique();
> 
>     let mut program_test = ProgramTest::new(
>         "helloworld", // Run the BPF version with `cargo test-bpf`
>         program_id,
>         processor!(process_instruction), // Run the native version with `cargo test`
>     );
> ```
> 
> but... this is nowhere to be found in case of anchor
> 
> ```
>         processor!(process_instruction), // Run the native version with `cargo test`
> ```
> 
> any ideas how to work with that?

Please don't derail issues with unrelated questions. You can use the generated entrypoint for rust tests.

**Comment by tomlinton** [comment]
@colinlgray Are you still looking at this?

**Comment by colinlgray** [comment]
I'm not. I didn't really get the intent of what this is supposed to be. I'm more of a front end developer and haven't done a lot of work on anchor. I thought that this would be a good place to jump in but I'm realizing that I don't know enough about anchor and the intent of this ticket to do it. Also I realized the CRA app template wouldn't be needed because it would be easy enough to use git to clone dapp-scaffold

**Comment by jamie-osec** [comment]
I'm going to close since I believe this is out of scope for what Anchor should provide
